src/processor.jl

```diff
@@ -1,19 +1,237 @@
+
+"""
+    get_parent(proc::Processor) -> Processor
+
+Returns the parent processor for `proc`. The ultimate parent processor is an
+`OSProc`. `Processor` subtypes should overload this to return their most
+direct parent.
+"""
+get_parent
+
+"""
+    move(from_proc::Processor, to_proc::Processor, x)
+
+Moves and/or converts `x` such that it's available and suitable for usage on
+the `to_proc` processor. This function can be overloaded by `Processor`
+subtypes to transport arguments and convert them to an appropriate form before
+being used for exection. This is additionally called on thunk results when
+moving back to `from_proc` before being serialized over the wire as needed.
+The default implementation breaks a single `move` call down into a sequence of
+`move` calls, and is not intended to be maximally efficient.
+"""
+function move(ctx, from_proc::Processor, to_proc::Processor, x)
+    @debug "Initiating generic move"
+    # Move to remote OSProc
+    @debug "(Remote) moving $parent_proc to $grandparent_proc"
+    root = get_parent_osproc(from_proc)
+    x, parent_proc = remotecall_fetch(move_to_osproc, root.pid, from_proc, x)
+
+    # Move to local OSProc
+    remote_proc = parent_proc
+    local_proc = OSProc()
+    @debug "(Network) moving $remote_proc to $local_proc"
+    x = move(ctx, remote_proc, local_proc, x)
+
+    # Move to to_proc
+    parent_proc = get_parent(to_proc)
+    path = Processor[to_proc, parent_proc]
+    while parent_proc != local_proc && !(parent_proc isa OSProc)
+        parent_proc = get_parent(parent_proc)
+        push!(path, parent_proc)
+    end
+    last_proc = local_proc
+    while !isempty(path)
+        next_proc = pop!(path)
+        @debug "(Local) moving $last_proc to $next_proc"
+        x = move(ctx, last_proc, next_proc, x)
+        last_proc = next_proc
+    end
+    return x
+end
+function get_parent_osproc(proc)
+    while !(proc isa OSProc)
+        proc = get_parent(proc)
+    end
+    proc
+end
+function move_to_osproc(parent_proc, x)
+    ctx = Context()
+    while !(parent_proc isa OSProc)
+        grandparent_proc = get_parent(parent_proc)
+        x = move(ctx, parent_proc, grandparent_proc, x)
+        parent_proc = grandparent_proc
+    end
+    return x, parent_proc
+end
+
+"""
+    OSProc <: Processor
+
+Julia CPU (OS) process, identified by Distributed pid. Executes thunks when
+threads or other "children" processors are not available, and/or the user has
+not opted to use those processors.
 """
 struct OSProc <: Processor
     pid::Int
+    attrs::Dict{Symbol,Any}
+    children::Vector{Processor}
+    queue::Vector{Processor}
+end
+const OSPROC_CACHE = Dict{Int,OSProc}()
+OSProc(pid::Int=myid()) = get!(OSPROC_CACHE, pid) do
+    remotecall_fetch(get_osproc, pid, pid)
+end
+function get_osproc(pid::Int)
+    proc = OSProc(pid, Dict{Symbol,Any}(), Processor[], Processor[])
+    for cb in PROCESSOR_CALLBACKS
+        try
+            child = cb(proc)
```

Comment:
`invokelatest`

Reply:
Good catch! Fixed and added a test for this :smile: